### PR TITLE
git: Clone HEAD should not force master. Fixes #363

### LINKED ISSUE #249: Error with single branch clone with non-master default branch
[All new GitHub repos][1] are created with a default branch of `main`, instead of the older `master`.

If I try to clone such a repo with the `SingleBranch` option enabled, I see an error like `couldn't find remote ref "refs/heads/master"`.

This code demonstrates the issue (the repo https://github.com/hairyhenderson/hitron_coda.git has a default branch of `main`):

```go
package main

import (
	"fmt"

	"github.com/go-git/go-billy/v5/memfs"
	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/storage/memory"
)

func main() {
	fs := memfs.New()
	storer := memory.NewStorage()

	opts := &git.CloneOptions{
		URL:          "https://github.com/hairyhenderson/hitron_coda.git",
		SingleBranch: true,
	}

	_, err := git.Clone(storer, fs, opts)
	if err != nil {
		fmt.Println(err)
	}
}
```

```console
$ go run main.go
couldn't find remote ref "refs/heads/master"
```

If I set `SingleBranch` to `false` it succeeds, but this isn't desirable for my use-case since I want to minimize what I'm pulling down.

The issue seems rooted in [this code](https://github.com/go-git/go-git/blob/v5.2.0/repository.go#L923), which defaults to the [`plumbing.Master`](https://github.com/go-git/go-git/blob/v5.2.0/plumbing/reference.go#L128) constant when `SingleBranch` is `true` and `ReferenceName` defaults to `HEAD`.

[1]: https://github.blog/changelog/2020-10-01-the-default-branch-for-newly-created-repositories-is-now-main/

### LINKED ISSUE #363: Cloning with options single branch and ref = HEAD is broken for repo with main base branch
Hey y'all ✌️🤠!

Repositories are shifting over to using `main` as their base branch. 
Today `go-git` does not work as expected when cloning a repo with main = head and the following options:
```Golang
&git.CloneOptions{
 	URL:          url,
 	Depth:        1,
 	SingleBranch: true,
 	ReferenceName: plumbing.HEAD,
 }
``` 
I _kind_ of dug around and noticed that `go-git` hardcodes to `plumbing.Master.Short()` here: 
https://github.com/go-git/go-git/blob/4ec1753b4e9324d455d3b55060020ce324e6ced2/repository.go#L905-L908

This is the git command equivalent: 
`git clone --single-branch --depth=1 {REPO_URL_WITH_MAIN_BRANCH_AS_BASE}`
The expected behavior is that it clones the repository with `HEAD` pointing to `main`. `go-git` returns an error stating `couldn't find remote ref "refs/heads/master"`.

Is this something that y'all are aware of or on y'all's road map?
I am more than happy to help contribute a fix! 
Just would need some guidance & more time to figure out how `config.RefSpec` works and why it's hardcoded to `plumbing.Master.Short()` in the switch statement above.

### PR DESCRIPTION
**Pull request by AriehSchneier**
Also fixes #249

### CODE REVIEW

File: repository_test.go [hunk 1]
```diff
@@ -1071,6 +1071,49 @@ func (s *RepositorySuite) testCloneSingleBranchAndNonHEADReference(c *C, ref str
...
 
+func (s *RepositorySuite) TestCloneSingleBranchHEADMain(c *C) {
+	r, _ := Init(memory.NewStorage(), nil)
+
+	head, err := r.Head()
+	c.Assert(err, Equals, plumbing.ErrReferenceNotFound)
+	c.Assert(head, IsNil)
+
+	err = r.clone(context.Background(), &CloneOptions{
+		URL:          "https://github.com/AriehSchneier/main.git",
```
Comment (AriehSchneier) [inline repository_test.go:1082]: @pjbgf I would appreciate if you created me a repo in [`git-fixtures`](https://github.com/git-fixtures) to use instead

**Comment by AriehSchneier** [comment]
Rebased

File: repository_test.go (hunk 1 above)
Comment (pjbgf) [reply repository_test.go:1130]: @mcuadros do you mind creating a new repository where default branch is `main` please?

Comment (AriehSchneier) [reply repository_test.go]: I found a fixture that has a `main` and no `master`. I forced pushed an update, the only difference is the clone URL is now from fixtures and the hash check is updated.

Comment (AriehSchneier) [reply repository_test.go]: @pjbgf I found a fixture that has a main and no master. I forced pushed an update, the only difference is the clone URL is now from fixtures and the hash check is updated.

**Review by pjbgf — APPROVED** [review]
@AriehSchneier great job, thanks for working on this. :bow:

**Pull request merged**